### PR TITLE
CrowdStrike falcon integration enhancement

### PR DESCRIPTION
**Pull request by ChanochShayner**
## XSOAR [Partner](https://xsoar.pan.dev/docs/partners/become-a-tech-partner)?
- [ ] The title must be in the following format: **[YOUR_PARTNER_ID] short description**.

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/25228

## Description
add fetch incidents and detections and commands to the integration

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 1]
```diff
@@ -1110,70 +1156,137 @@ def fetch_incidents():
         Fetches incident using the detections API
         :return: Fetched detections in incident format
     """
-    last_run = demisto.getLastRun()
-    # Get the last fetch time, if exists
-    last_fetch = last_run.get('first_behavior_time')
+    incidents = []  # type:List
+    incidents_or_detections = demisto.params().get('fetch_incidents_or_detections')
```
Comment (guykeller) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1160]: let's find a more descriptive name

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 2]
```diff
@@ -1110,70 +1156,137 @@ def fetch_incidents():
...
-            if len(detections_ids) == 1:
-                return incidents
+            # Limit the results to INCIDENTS_PER_FETCH`z
+            last_index_to_fetch = INCIDENTS_PER_FETCH + first_index_to_fetch
+            detections_ids = detections_ids[first_index_to_fetch:last_index_to_fetch]
+            raw_res = get_detections_entities(detections_ids)
+
+            if "resources" in raw_res:
+                raw_res['type'] = "detections"
+                for detection in raw_res.get("resources"):
```
Comment (guykeller) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1205]: why did you change these lines?

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 3]
```diff
@@ -1110,70 +1156,137 @@ def fetch_incidents():
...
+
+                    incidents.append(incident)
+
+            demisto.setLastRun({'first_behavior_time': last_fetch, 'last_detection_id': last_detection_id})
+
+    if 'incidents' in incidents_or_detections:
+
+        last_run = demisto.getLastRun()
+
+        last_fetch = last_run.get('first_behavior_incident_time')
```
Comment (guykeller) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1230]: what does this name mean?

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 4]
```diff
@@ -1110,70 +1156,137 @@ def fetch_incidents():
...
+
+        if last_fetch is None:
```
Comment (guykeller) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1232]:
```suggestion
        if not last_fetch:
```

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 5]
```diff
@@ -1110,70 +1156,137 @@ def fetch_incidents():
...
+            # make sure we do not fetch the same detection again.
+            if last_incident_id == incidents_ids[0]:
+                first_index_to_fetch = 1
+
+                # if this is the only detection - dont fetch.
+                if len(incidents_ids) == 1:
+                    return incidents
+
+            # if the first detection in this pull is different than the last detection fetched we bring it as well
+            else:
```
Comment (guykeller) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1259]: A lot of the code here and down the line seems to be duplicate.
Let's extract any duplicate logic to external functions

**Review by guykeller — CHANGES_REQUESTED** [review]

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 6]
```diff
@@ -861,6 +879,34 @@ def get_detections_entities(detections_ids):
     return detections_ids
 
 
+def get_fetch_incidents(last_created_timestamp=None, filter_arg=None):
+    endpoint_url = '/incidents/queries/incidents/v1'
+    params = {
+        'sort': 'modified_timestamp.asc'
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:885]: There is an option to sort the events in ascending / descending order

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py (hunk 2 above)
Comment (ChanochShayner) [reply Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1205]: The first line - we decided to add to each incident its type - incident or detection.
Second line - my mistake

File: Packs/CommonTypes/ReleaseNotes/1_9_0.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
+
+#### Incident Fields
+- **Account Name**
+- **Device Model**
+- **MD5**
```
Comment (ronykoz) [inline Packs/CommonTypes/ReleaseNotes/1_9_0.md:5]: md5, sha256 and i think all of the rest are existing fields, @David-BMS could you help out here please?

**Comment by ronykoz** [comment]
@ChanochShayner let us know once we can re-review it

File: Packs/CommonTypes/ReleaseNotes/1_9_6.md [hunk 8]
```diff
@@ -0,0 +1,9 @@
+#### Incident Fields
+- **Account Name**
+- **Device Model**
+- **MD5**
+- **Device External IP**
+- **SHA256**
+- **Assigned User**
+- **Device Local IP**
+- Added incident fields for CrowdStrikeFalcon
```
Comment (ronykoz) [inline Packs/CommonTypes/ReleaseNotes/1_9_6.md:9]: pretty sure you need a separate line for each of them

File: Packs/CrowdStrikeFalcon/Classifiers/classifier-CrowdStrike_Falcon_Incident_Classifier.json [hunk 9]
```diff
@@ -0,0 +1,14 @@
+{
+	"id": "CrowdStrike Falcon",
+	"name": "CrowdStrike Falcon Incident Classifier",
```
Comment (ronykoz) [inline Packs/CrowdStrikeFalcon/Classifiers/classifier-CrowdStrike_Falcon_Incident_Classifier.json:3]: let's update the description

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.yml [hunk 10]
```diff
@@ -46,43 +46,80 @@ configuration:
   name: proxy
   required: false
   type: 8
+- additionalinfo: Choose what to fetch - incidents or detections or both
+  defaultvalue: detections
+  display: Choose what to fetch
```
Comment (ronykoz) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.yml:51]: Fetch types

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.yml [hunk 11]
```diff
@@ -46,43 +46,80 @@ configuration:
...
+  hidden: false
+  name: fetch_incidents_or_detections
+  options:
+  - Detections
+  - Incidents
```
Comment (ronykoz) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.yml:56]: add:
`- Incidents and Detections`

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 12]
```diff
@@ -1118,61 +1166,110 @@ def timestamp_length_equalization(timestamp1, timestamp2):
...
+    incidents = []  # type:List
 
-    # Handle first time fetch, fetch incidents retroactively
-    if last_fetch is None:
-        last_fetch, _ = parse_date_range(FETCH_TIME, date_format='%Y-%m-%dT%H:%M:%SZ')
-    prev_fetch = last_fetch
+    fetch_incidents_or_detections = demisto.params().get('fetch_incidents_or_detections', '')
 
-    last_fetch_timestamp = int(parse(last_fetch).timestamp() * 1000)
+    if 'Detections' in fetch_incidents_or_detections or not fetch_incidents_or_detections:
```
Comment (ronykoz) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1189]: or not fetch_incidents_or_detections ?

**Review by ronykoz — CHANGES_REQUESTED** [review]
The image should state the name of the product
@guykeller Let's do a drill down to the code

**Review by guykeller — APPROVED** [review]

**Review by ronykoz — APPROVED** [review]

**Review by guykeller — APPROVED** [review]

**Comment by spearmin10** [comment]
No incidents is fetched in `Detections` mode.  Please fix LINE:1275 in `fetch_incidents()`
from
        return incidents
to
    return incidents
